src/js/azdata/azdataQueryPlan.js

```diff
@@ -1,3 +1,126 @@
+    }
+
+    checkInvariant() {
+        var last = new Point(0, 0);
+
+        for (var i = 0; i < this.layoutPoints.length; i++) {
+            if (last.x > this.layoutPoints[i].x || last.Y > this.layoutPoints[i].y) {
+                console.log('Graph layout failed.');
+                // do not throw exception, in case of failure we render graph incorrectly 
+            }
+
+            last = this.layoutPoints[i];
+        }
+    }
+
+    updateNodeLayout(xPosition, yPosition) {
+        this.checkInvariant();
+
+        // First cover edge cases
+
+        // Empty list
+        if (this.layoutPoints.length === 0) {
+            this.layoutPoints.push(new Point(xPosition, yPosition));
+            return;
+        }
+
+        // Single Element
+        if (this.layoutPoints.length === 1) {
+            if (xPosition < this.layoutPoints[0].x) {
+                this.layoutPoints.splice(0, 0, new Point(xPosition, yPosition));
+            }
+            else if (xPosition === this.layoutPoints[0].x) {
+                this.layoutPoints[0] = new Point(this.layoutPoints[0].x, Math.max(this.layoutPoints[0].y, yPosition));
+            }
+            else {
+                this.layoutPoints.push(new Point(xPosition, yPosition));
+            }
+            return;
+        }
+
+        // Insert Before First Element
+        if (xPosition < this.layoutPoints[0].x && yPosition < this.layoutPoints[0].y) {
+            this.layoutPoints.splice(0, 0, new Point(xPosition, yPosition));
+            return;
+        }
+
+        // Insert Last Element
+        if (this.layoutPoints[this.layoutPoints.length - 1].x < xPosition && yPosition > this.layoutPoints[this.layoutPoints.length - 1].y) {
+            this.layoutPoints.push(new Point(xPosition, yPosition));
+            return;
+        }
+
+        // Update Last Element
+        if (this.layoutPoints[this.layoutPoints.length - 1].x == xPosition) {
+            this.layoutPoints[this.layoutPoints.length - 1] = new Point(xPosition, Math.max(this.layoutPoints[this.layoutPoints.length - 1].y, yPosition));
+            return;
+        }
+
+        // Insert Point 
+
+        // First find insert index
+        var insertIndex = 0;
+        for (var i = 0; i < this.layoutPoints.length; i++) {
+            if (xPosition <= this.layoutPoints[i].x) {
+                insertIndex = i;
+                break;
+            }
+        }
+
+        // Perform Insert or Update.
+        if (xPosition == this.layoutPoints[insertIndex].x) {
+            this.layoutPoints[insertIndex] = new Point(xPosition, Math.max(this.layoutPoints[insertIndex].y, yPosition));
+        }
+        else {
+            this.layoutPoints.splice(insertIndex, 0, new Point(xPosition, yPosition));
+        }
+
+        // After we insert the point we need to remove following points if they have lower Y value.
+        var lastIndex = insertIndex;
+
+        while (lastIndex < this.layoutPoints.length) {
```

Comment:
NIT: `lastIndex` isn't referenced outside the while loop, so you can just use a for loop to iterate over the following layout points after inserting.